### PR TITLE
Ladders: Fix crash

### PR DESCRIPTION
**Pull request by mia-pi-git**
A user namechanging before the battle ended in any way would break this.

### CODE REVIEW

File: server/ladders-remote.ts [hunk 1]
```diff
@@ -70,14 +70,40 @@ export class LadderStore {
 		const formatid = this.formatid;
 		const p1 = Users.getExact(p1name);
 		const p2 = Users.getExact(p2name);
-		room.update();
-		room.send(`||Ladder updating...`);
-		const [data, error] = await LoginServer.request('ladderupdate', {
+		const p1id = p1 ? p1.id : toID(p1name);
+		const p2id = p2 ? p2.id : toID(p2name);
```
Comment (Zarel) [inline server/ladders-remote.ts:74]: Why not just always `toID(p1name)`?

File: server/ladders-remote.ts [hunk 2]
```diff
@@ -70,14 +70,40 @@ export class LadderStore {
...
+		room.addRaw(Utils.html`${p2name}'s rating: ${p2OldElo} &rarr; <strong>${p2NewElo}</strong><br />(${p2Reasons})`);
+
+		room.rated = Math.min(p1NewElo, p2NewElo);
+
+		if (p1) p1.mmrCache[formatid] = +p1NewElo;
+		if (p2) p2.mmrCache[formatid] = +p2NewElo;
+
+		room.update();
+
+
```
Comment (Zarel) [inline server/ladders-remote.ts:105]:
```suggestion

```

File: server/ladders-remote.ts (hunk 1 above)
Comment (mia-pi-git) [inline server/ladders-remote.ts:74]:
```suggestion
		const p1id = toID(p1name);
		const p2id = toID(p2name);
```

**Review by AnnikaCodes — APPROVED** [review]
Thanks!

File: server/ladders-remote.ts [hunk 3]
```diff
@@ -70,14 +70,38 @@ export class LadderStore {
...
+		const p1Act = (p1score > 0.9 ? `winning` : (p1score < 0.1 ? `losing` : `tying`));
+		let p1Reasons = `${p1NewElo - p1OldElo} for ${p1Act}`;
+		if (!p1Reasons.startsWith('-')) p1Reasons = '+' + p1Reasons;
+		room.addRaw(Utils.html`${p1name}'s rating: ${p1OldElo} &rarr; <strong>${p1NewElo}</strong><br />(${p1Reasons})`);
+
+		const p2Act = (p1score > 0.9 || p1score < 0 ? `losing` : (p1score < 0.1 ? `winning` : `tying`));
+		let p2Reasons = `${p2NewElo - p2OldElo} for ${p2Act}`;
+		if (!p2Reasons.startsWith('-'))	p2Reasons = '+' + p2Reasons;
+		room.addRaw(Utils.html`${p2name}'s rating: ${p2OldElo} &rarr; <strong>${p2NewElo}</strong><br />(${p2Reasons})`);
```
Comment (Zarel) [inline server/ladders-remote.ts:96]: You forgot to remove the old rating messages when you added these new messages in, so now messages are appearing twice.

File: server/ladders-remote.ts [hunk 4]
```diff
@@ -149,4 +173,34 @@ export class LadderStore {
 	static async visualizeAll(username: string) {
 		return [`<tr><td><strong>Please use the official client at play.pokemonshowdown.com</strong></td></tr>`];
 	}
+	/**
+	 * Calculates Elo based on a match result
+	 */
+	private calculateElo(previousUserElo: number, score: number, foeElo: number): number {
```
Comment (Zarel) [inline server/ladders-remote.ts:179]: This formula is wrong; it's generating different results from `ntbb-ladder.php`:

> Kennedy ⌐◼_◼'s rating: 1000 → 1064
> (+64 for winning)
> eaubot's rating: 1157 → 1128
> (-29 for losing)
> Kennedy ⌐◼_◼'s rating: 1000 → 1057
> (+57 for winning)
> eaubot's rating: 1157 → 1121
> (-36 for losing)

(gen8challengecup1v1 battle)

Comment (Zarel) [reply server/ladders-remote.ts:179]: Your duplicate result bug at least allowed us to catch this...

**Pull request merged**